README.md
# Description

Library of vault resources to handle secrets

# Requirements

## Platform:

* aix
* amazon
* centos
* fedora
* freebsd
* debian
* oracle
* mac_os_x
* redhat
* suse
* opensuse
* opensuseleap
* ubuntu
* windows
* zlinux

## Cookbooks:

*No dependencies defined*

# Attributes

*No attributes defined*

# Recipes

* vault_resources::default
* vault_resources::vault_certificate_example

# Resources

* [vault_certificate](#vault_certificate)

## vault_certificate

This resource allow to request certificates from a vault server

### Actions

- issue: Require the certificate Default action.

### Attribute Parameters

- common_name: The common_name of the certificate, taken fromt he resource name
- certificate_path: Where to write the resulting certificate files
- owner:
- mode:  Defaults to <code>"0600"</code>.
- group:
- alt_names: Array of alternative names (sans) to add into the certificate Defaults to <code>[]</code>.
- ip_sans: Array of alternative names in IP format to add into the certificate Defaults to <code>[]</code>.
- ttl_days: Validity of the certificate in days Defaults to <code>365</code>.
- reissue_within_days: Issue a new certificate if the current one is due to expire in less than this value in days Defaults to <code>15</code>.
- force_issue:  Defaults to <code>false</code>.
- vault_auth_method:  Defaults to <code>"token"</code>.
- vault_auth_credentials:  Defaults to <code>[]</code>.
- vault_client_options: Define the option to pass to vault client, could be empty to use environment variables. Defaults to <code>{}</code>.
- vault_role:  Defaults to <code>"pki/issue/webserver"</code>.

### Remarks

  The certificates files will be created under <certificate_path>/<common_name>

  - `vault_auth method` is one supported by [Vault::Authenticate](https://www.rubydoc.info/github/hashicorp/vault-ruby/Vault/Authenticate)
  - `vault_auth_credentials` is an array which should match the method parameters of `vault_auth_method`
  - `vault_client_options` is a hash for which keys should be in Vault::Configurable.keys, the token will be set by the auth method and can be avoided
  - `vaul_role` is the role in the pki backend to which requesting the certificate, must be of the form `<backend>/issue/<role>`

  You can omit the vault_client_options like address if you want to use the environment variables for vault.

### Examples

  ```
  vault_certificate 'test.example.com' do
    certificate_path '/tmp'
    alt_names ['test2.example.com', 'test3.example.com']
    ip_sans ['127.0.0.1', node['ipaddress']]
    ttl_days 30

    vault_auth_method 'token'
    vault_auth_credentials ['3VRLzAKbpU58Wgio1a5K8G1E']
    vault_client_options('server_url' => 'http://127.0.0.1:8200')
    vault_role 'pki/issue/webserver'
  end
  ```

# License and Maintainer

Maintainer:: Sous Chefs (<help@sous-chefs.org>)

Source:: https://github.com/sous-chefs/vault_resources

Issues:: https://github.com/sous-chefs/vault_resources/issues

License:: MIT


CHANGELOG.md
# vault_resources CHANGELOG

This file is used to list changes made in each version of the vault_resources cookbook.

# 0.1.0

Initial release.

- creation of the vault_certificate resource



spec/unit/recipes/default_spec.rb
#
# Cookbook:: vault_resources
# Spec:: default
#
# Copyright:: 2018, The Authors, All Rights Reserved.

require 'spec_helper'

describe 'vault_resources::default' do
  context 'When all attributes are default, on Ubuntu 16.04' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::ServerRunner.new(platform: 'ubuntu', version: '16.04')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end

  context 'When all attributes are default, on CentOS 7.4.1708' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::ServerRunner.new(platform: 'centos', version: '7.4.1708')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


test/integration/default/default_test.rb
# # encoding: utf-8

# Inspec test for recipe vault_resources::default

# The Inspec reference, with examples and extensive documentation, can be
# found at http://inspec.io/docs/reference/resources/

# This is an example test, replace it with your own test.
describe port(80), :skip do
  it { should_not be_listening }
end


spec/unit/recipes/vault_certificate_example_spec.rb
#
# Cookbook:: vault_resources
# Spec:: default
#
# Copyright:: 2018, The Authors, All Rights Reserved.

require 'spec_helper'

describe 'vault_resources::vault_certificate_example' do
  context 'When creating a certificate' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::ServerRunner.new(platform: 'ubuntu', version: '16.04')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


resources/pki_role.rb
require 'vault'

resource_name :vault_pki_intermediate

property :ttl, String, default: '2160h'
property :max_ttl, String, default: '8760h'
property :allow_localhost, [TrueClass, FalseClass], default: true
property :allowed_domains, Array, default: []
property :allow_bare_domains, [TrueClass, FalseClass], default: false
property :allow_subdomains, [TrueClass, FalseClass], default: false
property :allow_glob_domains, [TrueClass, FalseClass], default: false
property :allow_any_name,[TrueClass, FalseClass], default: false
property :enforce_hostnames, [TrueClass, FalseClass], default: true
property :allow_ip_sans, [TrueClass, FalseClass], default: false
property :allowed_uri_sans, String, default: ""
property :server_flag, [TrueClass, FalseClass], default: true
property :client_flag, [TrueClass, FalseClass],default: true
property :code_signing_flag, [TrueClass, FalseClass], default: false
property :email_protection_flag, [TrueClass, FalseClass], default: false
property :key_type, %w(rsa ec), default: 'rsa'
property :key_bits, Integer, default: 2048
property :key_usage, Array, default: %w(DigitalSignature KeyAgreement KeyEncipherment)
property :ext_key_usage, Array, default: []
property :use_csr_common_name, [TrueClass, FalseClass], default: true
property :use_csr_sans, [TrueClass, FalseClass], default: true
property :ou, Array, default: []
property :organization, Array, default: []
property :country, Array, default: []
property :locality, Array, default: []
property :province, Array, default: []
property :street_address, Array, default: []
property :postal_code, Array, default: []
property :generate_lease, [TrueClass, FalseClass], default: false
property :no_store, [TrueClass, FalseClass], default: false
property :require_cn, [TrueClass, FalseClass], default: true
property :policy_identifiers, Array,  default: []
property :basic_constraints_valid_for_non_ca, [TrueClass, FalseClass], default: false
property :not_before_duration, String, default: '30s'

property :vault_backend, String, default: 'pki', desired_state: false
property :vault_auth_method, String, default: 'token', desired_state: false, callbacks: {
    "should be one of Vault::Authenticate methods: #{Vault::Authenticate.instance_methods(false)}" => lambda do |m|
      Vault.auth.respond_to?(m)
    end,
}
property :vault_auth_credentials, Array, desired_state: false, default: [], sensitive: true
#<> @property vault_client_options Define the option to pass to vault client, could be empty to use environment variables.
property :vault_client_options, Hash, desired_state: false, default: {}, callbacks: {
    "options should only include valid keys: #{Vault::Configurable.keys}" => lambda do |v|
      (v.keys.map { |k| k.is_a?(String) ? k.to_sym : k } - Vault::Configurable.keys).empty?
    end,
    'address should be a valid url' => lambda do |v|
      v.empty? || URI.parse(v['address'])
    end,
}
#<> @property vault_role where to mount this pki backend in vault.
property :vault_role, String, name_property: true

load_current_value do |desired|
  vault_auth
  begin
    role = @vault.with_retries do |attempts, error|
      Chef::Log.info "Received exception #{error.class} from Vault - attempt #{attempts}" unless attempts == 0
      @vault.logical.read(
          "/#{vault_backend}/roles/#{vault_role}",

      )
    end
    role.data.each do |k,v|
      send(k,v)
    end
  rescue Vault::HTTPError => e
    current_value_does_not_exist!
  end
end

action :create do
  converge_if_changed do
    begin
      @vault.with_retries do |attempts, error|
        Chef::Log.info "Received exception #{error.class} from Vault - attempt #{attempts}" unless attempts == 0
        @vault.logical.write(
            "/#{vault_backend}/roles/#{vault_role}",
            ttl: new_resource.ttl,
            max_ttl: new_resource.max_ttl,
            allow_localhost: new_resource.allow_localhost,
            allowed_domains: new_resource.allowed_domains,
            allow_bare_domains: new_resource.allow_bare_domains,
            allow_subdomains: new_resource.allow_subdomains,
            allow_glob_domains: new_resource.allow_glob_domains,
            allow_any_name: new_resource.allow_any_name,
            enforce_hostnames: new_resource.enforce_hostnames,
            allow_ip_sans: new_resource.allow_ip_sans,
            allowed_uri_sans: new_resource.allowed_uri_sans,
            server_flag: new_resource.server_flag,
            client_flag: new_resource.client_flag,
            code_signing_flag: new_resource.code_signing_flag,
            email_protection_flag: new_resource.email_protection_flag,
            key_type: new_resource.key_type,
            key_bits: new_resource.key_bits,
            key_usage: new_resource.key_usage,
            ext_key_usage: new_resource.ext_key_usage,
            use_csr_common_name: new_resource.use_csr_common_name,
            use_csr_sans: new_resource.use_csr_sans,
            ou: new_resource.ou,
            organization: new_resource.organization,
            country: new_resource.country,
            locality: new_resource.locality,
            province: new_resource.province,
            street_address: new_resource.street_address,
            postal_code: new_resource.postal_code,
            generate_lease: new_resource.generate_lease,
            no_store: new_resource.no_store,
            require_cn: new_resource.require_cn,
            policy_identifiers: new_resource.policy_identifiers,
            basic_constraints_valid_for_non_ca: new_resource.basic_constraints_valid_for_non_ca,
            not_before_duration: new_resource.not_before_duration
        )
      end
    rescue Vault::HTTPError => e
      message = "Failed to create pki role - #{new_resource.vault_role}.\n#{e.message}"
      Chef::Log.fatal message
    end
  end
end

action :delete do
  converge_by do
    begin
      @vault.with_retries do |attempts, error|
        Chef::Log.info "Received exception #{error.class} from Vault - attempt #{attempts}" unless attempts == 0
        @vault.logical.write(
            "/#{vault_backend}/roles/#{vault_role}"
        )
      end
    rescue Vault::HTTPError => e
      message = "Failed to delete pki role - #{new_resource.vault_role}.\n#{e.message}"
      Chef::Log.fatal message
    end
  end
end

action_class do
  def vault_auth
    @vault = Vault::Client.new(new_resource.vault_client_options)
    @vault.auth.send new_resource.vault_auth_method, *new_resource.vault_auth_credentials
  end
end

resources/pki_root.rb
# rubocop:disable Layout/LeadingCommentSpace, Style/BlockComments
require 'vault'

resource_name :vault_pki



property :common_name, String, name_property: true
property :destination_path, String, default: '/tmp', desired_state: false
#<> @property type internal or exported, if exported the private key will be written with the certificate in `destination_path/common_name`
property :type, ['internal','exported'], default: 'internal', desired_state: false
#<> @property alt_names Array of alternative names (sans) to add into the certificate
property :alt_names, Array, default: []
property :ip_sans, Array, default: []
property :uri_sans, Array, default: []
property :other_sans, Array, default: []
#<> @property ttl_days Validity of the CA certificate in days
property :ttl_days, Integer, default: 365
property :format, %w(pem der pem_bundle), default: 'pem', desired_state: false
property :private_key_format, %w(der pkcs8), desired_state: false
property :key_type, %w(rsa ec), default: 'rsa', desired_state: false
property :key_bits, Integer, default: 2048, desired_state: false
property :max_path_length, Integer, default: -1, desired_state: false
property :exclude_cn_from_sans, [true, false], default: false, desired_state: false
property :permitted_dns_domains, Array, default: [], desired_state: false
property :ou, String, default: ''
property :organization, String, default: ''
property :country, String, default: ''
property :locality, String, default: ''
property :province, String, default: ''
property :street_address, String, default: ''
property :postal_code, String, default: ''
# Files properties (when exported)
property :owner, String
property :mode, String, default: '0600'
property :group, String
# Vault properties
#<> @property reissue_within_days Issue a new CA certificate if the current one is due to expire in less than this value in days
property :reissue_within_days, Integer, default: 90, desired_state: false
property :force_issue, [true, false], default: false
property :vault_auth_method, String, default: 'token', desired_state: false, callbacks: {
    "should be one of Vault::Authenticate methods: #{Vault::Authenticate.instance_methods(false)}" => lambda do |m|
      Vault.auth.respond_to?(m)
    end,
}
property :vault_auth_credentials, Array, desired_state: false, default: [], sensitive: true
#<> @property vault_client_options Define the option to pass to vault client, could be empty to use environment variables.
property :vault_client_options, Hash, desired_state: false, default: {}, callbacks: {
    "options should only include valid keys: #{Vault::Configurable.keys}" => lambda do |v|
      (v.keys.map { |k| k.is_a?(String) ? k.to_sym : k } - Vault::Configurable.keys).empty?
    end,
    'address should be a valid url' => lambda do |v|
      v.empty? || URI.parse(v['address'])
    end,
}
#<> @property vault_role where to mount this pki backend in vault.
property :vault_role, String, default: 'pki', desired_state: false

load_current_value do |desired|
  cert_path = "#{desired.certificate_path}/#{desired.common_name}/certificate.pem"
  vault_auth
  begin
    current_cert = @vault.with_retries do |attempts, error|
      Chef::Log.info "Received exception #{error.class} from Vault - attempt #{attempts}" unless attempts == 0
      @vault.logical.read(
          "/#{new_resource.vault_role}/ca"
      )
    end
  rescue Vault::HTTPError => e
    current_cert = nil
  end
  begin
    urls = @vault.with_retries do |attempts, error|
      Chef::Log.info "Received exception #{error.class} from Vault - attempt #{attempts}" unless attempts == 0
      @vault.logical.read(
          "/#{new_resource.vault_role}/config/urls"
      )
    end
    unless urls.nil?
      issuing_certificates urls.data['issuing_certificates']
      crl_distribution_points urls.data['crl_distribution_points']
      ocsp_servers urls.data['ocsp_servers']
    end
  rescue Vault::HTTPError
    issuing_certificates ''
    crl_distribution_points ''
    ocsp_servers ''
  end
  if ::File.exist?(cert_path)
    current_cert = OpenSSL::X509::Certificate.new ::File.read cert_path
  end
  if current_cert != nil
    cert = OpenSSL::X509::Certificate.new ::File.read cert_path
    cert_common_name = cert.subject.to_s(OpenSSL::X509::Name::COMPAT).split('=')[1]
    common_name cert_common_name
    current_alt_names = []
    current_ip_sans = []
    current_uri_sans = []
    current_other_sans = []
    cert.issuer.to_a.each do |entry|
      case entry[0]
      when 'C'
        country entry[1]
      when 'ST'
        province entry[1]
      when 'L'
        locality entry[1]
      when 'O'
        organization entry[1]
      when 'OU'
        ou entry[1]
      when 'street'
        street_address entry[1]
      when 'postalCode'
        postal_code entry[1]
      end
    end
    cert.extensions.each do |ext|
      next unless ext.oid().to_s == 'subjectAltName'
      # "subjectAltName = DNS:test.wherefor.com, DNS:test2.wherefor.com, DNS:test3.wherefor.com, IP Address:127.0.0.1, IP Address:127.0.0.2"
      ext.to_s.split('=')[1].split(',').each do |name|
        type, value = name.split(':',1)
        case type.strip
        when 'DNS'
          current_alt_names.push(value)
        when 'IP Address'
          current_ip_sans.push(value)
        when 'URI'
          current_uri_sans.push(value)
        when 'otherName'
          current_other_sans.push(value)
        end
        alt_names current_alt_names.sort - [desired.common_name, cert_common_name]
        ip_sans current_ip_sans.sort
        uri_sans current_uri_sans.sort
        other_sans current_other_sans.sort
      end
    end
  else
    common_name ''
  end
end

action :issue do
  issuing, reason = should_issue?
  vault_auth
  if issuing
    converge_by reason do
      issue_ca
    end
  else
    converge_if_changed :common_name, :alt_names, :ip_sans do
      issue_ca
    end
  end
end

action_class do
  def should_issue?
    cert_path = "#{new_resource.destination_path}/#{new_resource.common_name}/certificate.pem"
    return [true, 'issuing because :force_issue is true'] if new_resource.force_issue
    return [true, "issuing because no cert at :certificate_path (#{cert_path}) exists"] unless ::File.exist?(cert_path)
    cert = OpenSSL::X509::Certificate.new ::File.read cert_path
    not_after = cert.not_after
    days_until_expiry = (not_after - Time.now.utc()) / (60 * 60 * 24)
    return [true, "reissuing because cert is set to expire within :reissue_within_days (#{new_resource.reissue_within_days}) (expires in #{days_until_expiry})"] if days_until_expiry < new_resource.reissue_within_days
    [false, '']
  end

  def vault_auth
    @vault = Vault::Client.new(new_resource.vault_client_options)
    @vault.auth.send new_resource.vault_auth_method, *new_resource.vault_auth_credentials
  end

  def issue_ca
    begin
      secret = @vault.with_retries do |attempts, error|
        Chef::Log.info "Received exception #{error.class} from Vault - attempt #{attempts}" unless attempts == 0
        @vault.logical.write(
            "/#{new_resource.vault_role}/root/generate/#{new_resource.type}",
            common_name: new_resource.common_name,
            alt_names: new_resource.alt_names.join(',').strip().chomp(','),
            ip_sans: new_resource.ip_sans.sort.join(',').strip().chomp(','),
            uri_sans: new_resource.uri_sans.sort.join(',').strip().chomp(','),
            other_sans: new_resource.other_sans.sort.join(',').strip().chomp(','),
            ttl: (new_resource.ttl_days * 24).to_s + 'h',
            format: new_resource.format,
            private_key_format: new_resource.private_key_format,
            key_type: new_resource.key_type,
            key_bits: new_resource.key_bits,
            max_path_length: new_resource.max_path_length,
            exclude_cn_from_sans: new_resource.exclude_cn_from_sans,
            permitted_dns_domains: new_resource.permitted_dns_domains,
            ou: new_resource.ou,
            organization: new_resource.organization,
            country: new_resource.country,
            locality: new_resource.locality,
            province: new_resource.province,
            street_address: new_resource.street_address,
            postal_code: new_resource.postal_code,
        )
      end
    rescue Vault::HTTPError => e
      message = "Failed to create CA cert - #{new_resource.common_name}.\n#{e.message}"
      Chef::Log.fatal message
    end
    if secret.nil?
      message = "Could not create CA cert - #{new_resource.common_name}, no data retrieved"
      Chef::Log.fatal message
      return
    end
    Chef::Log.warn secret.inspect
    converge_if_changed :issuing_certificatesn, :crl_distribution_points, :ocsp_servers do
      begin
        @vault.with_retries do |attempts, error|
          Chef::Log.info "Received exception #{error.class} from Vault - attempt #{attempts}" unless attempts == 0
          @vault.logical.write(
              "/#{new_resource.vault_role}/config/urls",
              issuing_certificates: new_resource.ca_url,
              crl_distribution_points: new_resource.crl_url,
              ocsp_servers: new_resource.ocsp_servers
          )
        end
      rescue Vault::HTTPError => e
        message = "Failed to set urls - #{new_resource.common_name}.\n#{e.message}"
        Chef::Log.fatal message
      end
    end
    return unless new_resource.type == 'exported'
    directory "#{new_resource.destination_path}/#{new_resource.common_name}" do
      owner new_resource.owner
      group new_resource.group
      mode new_resource.mode
      recursive true
    end
    secret.data.keys.each do |f|
      file "#{new_resource.certificate_path.chomp('/')}/#{new_resource.common_name}/#{f}.pem" do
        owner new_resource.owner
        group new_resource.group
        mode new_resource.mode
        content secret.data[f].to_s
        sensitive true
      end
    end
  end
end

resources/certificate.rb
# Work derived from:
# WhereTo vault_pki  https://github.com/wherefortravel/vault_pki
# sous-chef vault cookbook secret https://github.com/sous-chefs/vault

# rubocop:disable Layout/LeadingCommentSpace, Style/BlockComments
=begin
#<
This resource allow to request certificates from a vault server

@action issue Require the certificate

@section Remarks

  The certificates files will be created under <certificate_path>/<common_name>

  - `vault_auth method` is one supported by [Vault::Authenticate](https://www.rubydoc.info/github/hashicorp/vault-ruby/Vault/Authenticate)
  - `vault_auth_credentials` is an array which should match the method parameters of `vault_auth_method`
  - `vault_client_options` is a hash for which keys should be in Vault::Configurable.keys, the token will be set by the auth method and can be avoided
  - `vaul_role` is the role in the pki backend to which requesting the certificate, must be of the form `<backend>/issue/<role>`

  You can omit the vault_client_options like address if you want to use the environment variables for vault.

@section Examples

  ```
  vault_certificate 'test.example.com' do
    certificate_path '/tmp'
    alt_names ['test2.example.com', 'test3.example.com']
    ip_sans ['127.0.0.1', node['ipaddress']]
    ttl_days 30

    vault_auth_method 'token'
    vault_auth_credentials ['3VRLzAKbpU58Wgio1a5K8G1E']
    vault_client_options('server_url' => 'http://127.0.0.1:8200')
    vault_role 'pki/issue/webserver'
  end
  ```
#>
=end
require 'vault'

resource_name :vault_certificate

#<> @property common_name The common_name of the certificate, taken fromt he resource name
property :common_name, String, name_property: true
#<> @property certificate_path Where to write the resulting certificate files
property :certificate_path, String, required: true
property :owner, String
property :mode, String, default: '0600'
property :group, String
#<> @property alt_names Array of alternative names (sans) to add into the certificate
property :alt_names, Array, default: []
#<> @property ip_sans Array of alternative names in IP format to add into the certificate
property :ip_sans, Array, default: []
#<> @property ttl_days Validity of the certificate in days
property :ttl_days, Integer, default: 365
#<> @property reissue_within_days Issue a new certificate if the current one is due to expire in less than this value in days
property :reissue_within_days, Integer, default: 15, desired_state: false
property :force_issue, [true, false], default: false
property :vault_auth_method, String, default: 'token', desired_state: false, callbacks: {
  "should be one of Vault::Authenticate methods: #{Vault::Authenticate.instance_methods(false)}" => lambda do |m|
    Vault.auth.respond_to?(m)
  end,
}
property :vault_auth_credentials, Array, desired_state: false, default: [], sensitive: true
#<> @property vault_client_options Define the option to pass to vault client, could be empty to use environment variables.
property :vault_client_options, Hash, desired_state: false, default: {}, callbacks: {
  "options should only include valid keys: #{Vault::Configurable.keys}" => lambda do |v|
    (v.keys.map { |k| k.is_a?(String) ? k.to_sym : k } - Vault::Configurable.keys).empty?
  end,
  'address should be a valid url' => lambda do |v|
    v.empty? || URI.parse(v['address'])
  end,
}
property :vault_role, String, default: 'pki/issue/webserver', desired_state: false, callbacks: {
  'must be a issue endpoint' => lambda do |r|
    !(r =~ %r{.+/issue/.+}).nil?
  end,
}

load_current_value do |desired|
  cert_path = "#{desired.certificate_path}/#{desired.common_name}/certificate.pem"
  if ::File.exist?(cert_path)
    cert = OpenSSL::X509::Certificate.new ::File.read cert_path
    cert_common_name = cert.subject.to_s(OpenSSL::X509::Name::COMPAT).split('=')[1]
    common_name cert_common_name
    current_alt_names = []
    current_ip_sans = []
    cert.extensions.each do |ext|
      next unless ext.oid().to_s == 'subjectAltName'
      # "subjectAltName = DNS:test.wherefor.com, DNS:test2.wherefor.com, DNS:test3.wherefor.com, IP Address:127.0.0.1, IP Address:127.0.0.2"
      ext.to_s.split('=')[1].split(',').each do |name|
        type = name.split(':')[0]
        value = name.split(':')[1]
        if type.strip == 'DNS'
          current_alt_names.push(value)
        elsif type.strip == 'IP Address'
          current_ip_sans.push(value)
        end
        alt_names current_alt_names.sort - [desired.common_name, cert_common_name]
        ip_sans current_ip_sans.sort
      end
    end
  else
    common_name ''
  end
end

action :issue do
  issuing, reason = should_issue?
  vault_auth
  if issuing
    converge_by reason do
      issue_cert
    end
  else
    converge_if_changed :common_name, :alt_names, :ip_sans do
      issue_cert
    end
  end
end

action_class do
  def should_issue?
    cert_path = "#{new_resource.certificate_path}/#{new_resource.common_name}/certificate.pem"
    return [true, 'issuing because :force_issue is true'] if new_resource.force_issue
    return [true, "issuing because no cert at :certificate_path (#{cert_path}) exists"] unless ::File.exist?(cert_path)
    cert = OpenSSL::X509::Certificate.new ::File.read cert_path
    not_after = cert.not_after
    days_until_expiry = (not_after - Time.now.utc()) / (60 * 60 * 24)
    return [true, "reissuing because cert is set to expire within :reissue_within_days (#{new_resource.reissue_within_days}) (expires in #{days_until_expiry})"] if days_until_expiry < new_resource.reissue_within_days
    [false, '']
  end

  def vault_auth
    @vault = Vault::Client.new(new_resource.vault_client_options)
    @vault.auth.send new_resource.vault_auth_method, *new_resource.vault_auth_credentials
  end

  def issue_cert
    begin
      secret = @vault.with_retries do |attempts, error|
        Chef::Log.info "Received exception #{error.class} from Vault - attempt #{attempts}" unless attempts == 0
        @vault.logical.write(
          new_resource.vault_role,
          common_name: new_resource.common_name,
          alt_names: new_resource.alt_names.join(',').strip().chomp(','),
          ip_sans: new_resource.ip_sans.sort.join(',').strip().chomp(','),
          ttl: (new_resource.ttl_days * 24).to_s + 'h'
        )
      end
    rescue Vault::HTTPError => e
      message = "Failed to create cert - #{new_resource.common_name}.\n#{e.message}"
      Chef::Log.fatal message
    end
    if secret.nil?
      message = "Could not create cert - #{new_resource.common_name}"
      Chef::Log.fatal message
      return
    end
    Chef::Log.warn secret.inspect
    directory "#{new_resource.certificate_path}/#{new_resource.common_name}" do
      owner new_resource.owner
      group new_resource.group
      mode new_resource.mode
      recursive true
    end
    secret.data.keys.each do |f|
      file "#{new_resource.certificate_path.chomp('/')}/#{new_resource.common_name}/#{f}.pem" do
        owner new_resource.owner
        group new_resource.group
        mode new_resource.mode
        content secret.data[f].to_s
        sensitive true
      end
    end
  end
end
